### PR TITLE
created Secondary Billing contact, Tertiary billing contact roles in edit client information page for storing clients Emails

### PR DESCRIPTION
**Pull request by abhishekkhanduri**
#1895, #1804

### Description
Created Secondary Billing Contact Role for storing cc Emails
Created Tertiary Billing Contact Role for storing bcc Emails

### Checklist:

- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: Modules/Client/Config/config.php [hunk 1]
```diff
@@ -25,6 +25,13 @@
         ],
     ],
 
+    'client-roles'=>[
+        'Primary-billing-contact' => 'billing-contact',
```
Comment (mohitgusain) [inline Modules/Client/Config/config.php:29]:
```suggestion
        'primary-billing-contact' => 'billing-contact',
```

File: Modules/Client/Config/config.php [hunk 2]
```diff
@@ -25,6 +25,13 @@
...
+        'General-Point-of-contact' => 'general-contact',
```
Comment (mohitgusain) [inline Modules/Client/Config/config.php:30]: can you also format this and the below line as mentioned above?

File: Modules/Client/Entities/Client.php [hunk 3]
```diff
@@ -83,7 +83,17 @@ public function contactPersons()
 
     public function getBillingContactAttribute()
     {
-        return $this->contactPersons()->where('type', 'billing-contact')->first();
+        return $this->contactPersons()->where('type', config('client.client-roles.Primary-billing-contact'))->first();
```
Comment (mohitgusain) [inline Modules/Client/Entities/Client.php:86]: Change all the query after changing the config values

File: Modules/Invoice/Resources/views/modals/payment-received.blade.php [hunk 4]
```diff
@@ -1,3 +1,22 @@
+@php
+if(optional($invoice->client->tertiary_contact)->first() != null) {
+    $bccEmails='';
+    foreach($invoice->client->tertiary_contact as $bccEmail)
+    {$bccEmails .= ",$bccEmail->email";}
```
Comment (mohitgusain) [inline Modules/Invoice/Resources/views/modals/payment-received.blade.php:5]: Check once the formatting

File: Modules/Invoice/Resources/views/modals/payment-received.blade.php [hunk 5]
```diff
@@ -1,3 +1,22 @@
...
+    $bccEmails='';
+    foreach($invoice->client->tertiary_contact as $bccEmail)
+    {$bccEmails .= ",$bccEmail->email";} 
+    $bccEmails = substr($bccEmails, 1);
+} else {
+    $bccEmails = null;
+}
+if(optional($invoice->client->secondary_contact)->first() != null) {
+    $ccEmails='';
+    foreach($invoice->client->secondary_contact as $ccEmail)
```
Comment (mohitgusain) [inline Modules/Invoice/Resources/views/modals/payment-received.blade.php:12]: Check once this code is working or not

**Review by mohitgusain — COMMENTED** [review]
Please check the comments

File: Modules/Invoice/Resources/views/modals/payment-received.blade.php (hunk 5 above)
Comment (abhishekkhanduri) [reply Modules/Invoice/Resources/views/modals/payment-received.blade.php:14]: I have checked, the code is working fine

File: Modules/Invoice/Resources/views/modals/payment-received.blade.php [hunk 6]
```diff
@@ -1,3 +1,24 @@
...
+}
+if(optional($invoice->client->secondary_contact)->first() != null) {
+    $ccEmails='';
+    foreach($invoice->client->secondary_contact as $ccEmail) {
+        $ccEmails .= ",$ccEmail->email";
+    } 
+    $ccEmails = substr($ccEmails, 1);
+} else {
+    $ccEmails = null;
+}
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/modals/payment-received.blade.php:19]: can we send these 2 variables from service. We should not perform logics inside blade file.

File: Modules/Invoice/Resources/views/index.blade.php [hunk 7]
```diff
@@ -87,6 +87,23 @@
...
+                                $bccEmails = null;
+                            }
+                            if(optional($invoice->client->secondary_contact)->first() != null) {
+                                $ccEmails=config('invoice.mail.send-invoice.email');
+                                foreach($invoice->client->secondary_contact as $ccEmail) {
+                                    $ccEmails .= ",$ccEmail->email";
+                                }
+                            } else {
+                                $ccEmails = null;
+                            }
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/index.blade.php:106]: Lets create a getAttribute inside Module to access the bcc and cc emails

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
minor suggestions. Please have a look

File: Modules/Invoice/Entities/Invoice.php [hunk 8]
```diff
@@ -189,4 +189,31 @@ public function getTotalAmountAttribute()
...
+        if (optional($this->client->tertiary_contact)->first() != null) {
+            $bccEmails = '';
+            foreach ($this->client->tertiary_contact as $bccEmail) {
+                $bccEmails .= ",$bccEmail->email";
+            }
+            $bccEmails = substr($bccEmails, 1);
+        }
+
+        return $bccEmails;
+    }
```
Comment (GauravGusain98) [inline Modules/Invoice/Entities/Invoice.php:218]: We don't want this attribute, AS we can directly use like this `$invoice->client->bcc_emails`

File: Modules/Client/Entities/Client.php [hunk 9]
```diff
@@ -239,4 +249,31 @@ public function getClientMonthEndDateAttribute()
...
+        return $ccEmails;
+    }
+
+    public function getBccEmailsAttribute()
+    {
+        $bccEmails = null;
+        if (optional($this->tertiary_contact)->first() != null) {
+            $bccEmails = '';
+            foreach ($this->tertiary_contact as $bccEmail) {
+                $bccEmails .= ",$bccEmail->email";
```
Comment (GauravGusain98) [inline Modules/Client/Entities/Client.php:272]: Not sure how this is working correct, as you have passed variable inside a string.
```suggestion
                $bccEmails .= $tertiaryContact->email . ",";
```

File: Modules/Invoice/Resources/views/modals/payment-received.blade.php [hunk 10]
```diff
@@ -20,15 +20,18 @@ class="form-control" value="{{ config('invoice.mail.send-invoice.email') }}" req
...
+                    </div>
                     <div class="form-group col-md-12">
                         <label class="leading-none" for="cc">
                             {{ __('CC') }} 
                             <span data-toggle="tooltip" data-placement="right" title="Comma separated emails">
                                 <i class="fa fa-question-circle"></i>
                             </span>
                         </label>
-                        <input type="text" name="cc" id="cc" class="form-control" value="{{ config('invoice.mail.send-invoice.email') }}">
+                        <input type="text" name="cc" id="cc" class="form-control" value="{{$invoice->cc_emails == null ? config('invoice.mail.send-invoice.email') : config('invoice.mail.send-invoice.email') .','. $invoice->cc_emails}}">
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/modals/payment-received.blade.php:34]:
```suggestion
                        <input type="text" name="cc" id="cc" class="form-control" value="{{ config('invoice.mail.send-invoice.email') . ($invoice->cc_emails == null ?: ','. $invoice->cc_emails) }}">
```

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Left some comments, please check

**Review by GauravGusain98 — APPROVED** [review]
Looks fine

**Pull request merged**